## Exercises 1

In [3]:
from env import host, user, password

In [13]:
url = f'mysql+pymysql://{user}:{password}@{host}/employees'

In [14]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

In [58]:
import pandas as pd
import numpy as np
from datetime import date

In [7]:
def get_db_url(user, host, password, database):
    
    return f'mysql+pymysql://{user}:{password}@{host}/{database}'

In [8]:
url2 = get_db_url(user, host, password, 'employees')

In [23]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url2)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?

In [9]:
# NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:msql.pymysql
# used typo mysql --> mysql
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url2)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

b. Intentionally make an error in your SQL query. What does the error message look like?

In [10]:
pd.read_sql('SELECT * FROM employe LIMIT 5 OFFSET 50', url2)
# ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employe' doesn't exist")

ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employe' doesn't exist")
[SQL: SELECT * FROM employe LIMIT 5 OFFSET 50]
(Background on this error at: https://sqlalche.me/e/14/f405)

Read the employees and titles tables into two separate DataFrames.

In [12]:
url_emp = get_db_url(user, host, password, 'employees')

In [17]:
sql_tbl_emp = '''
SELECT *
FROM employees
'''

employees = pd.read_sql(sql_tbl_emp, url_emp)

In [18]:
employees.head(5)

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [19]:
sql_tbl_title = '''
SELECT *
FROM titles
'''

titles = pd.read_sql(sql_tbl_title, url_emp)

In [20]:
titles.head(5)

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

How many rows and columns do you have in each DataFrame? Is that what you expected?

In [22]:
employees.info() # 300024 rows, 6 columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [23]:
titles.info() # 443308 rows, 4 columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


Display the summary statistics for each DataFrame.

In [26]:
employees.describe(), titles.describe()

(              emp_no
 count  300024.000000
 mean   253321.763392
 std    161828.235540
 min     10001.000000
 25%     85006.750000
 50%    249987.500000
 75%    424993.250000
 max    499999.000000,
               emp_no
 count  443308.000000
 mean   253075.034430
 std    161853.292613
 min     10001.000000
 25%     84855.750000
 50%    249847.500000
 75%    424891.250000
 max    499999.000000)

How many unique titles are in the titles DataFrame? 7 

In [32]:
titles['title'].unique().size

7

What is the oldest date in the to_date column?

In [55]:
titles['to_date'].min()

datetime.date

What is the most recent date in the to_date column? 2002-08-01

In [78]:
titles[titles['to_date'] < date(9999, 1, 1)].max()

emp_no                 499998
title        Technique Leader
from_date          2002-06-30
to_date            2002-08-01
dtype: object

## Excerise II

Copy the users and roles DataFrames from the examples above.

In [79]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [81]:
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

What is the result of using a right join on the DataFrames?

In [120]:
users.merge(roles, how='right', right_on='id', left_on='role_id')

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0     1      admin
1   2.0    joe      2.0     2     author
2   3.0  sally      3.0     3   reviewer
3   4.0   adam      3.0     3   reviewer
4   NaN    NaN      NaN     4  commenter

What is the result of using an outer join on the DataFrames?

In [121]:
users.merge(roles, how='outer', left_on='role_id', right_on='id')

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0   1.0      admin
1   2.0    joe      2.0   2.0     author
2   3.0  sally      3.0   3.0   reviewer
3   4.0   adam      3.0   3.0   reviewer
4   5.0   jane      NaN   NaN        NaN
5   6.0   mike      NaN   NaN        NaN
6   NaN    NaN      NaN   4.0  commenter

What happens if you drop the foreign keys from the DataFrames and try to merge them?

In [125]:
users.drop('role_id', axis=1).merge(roles, how='right') # --> vertical merge

Empty DataFrame
Columns: [id, name, role_id]
Index: []

Load the mpg dataset from PyDataset.

In [132]:
from pydataset import data

Output and read the documentation for the mpg dataset.

In [133]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [134]:
mpg = data('mpg')

How many rows and columns are in the dataset? 234 rows, 11 columns

In [135]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


Check out your column names and perform any cleanup you may want on them.

In [138]:
mpg.columns

Index(['manufacturer', 'model', 'displ', 'year', 'cyl', 'trans', 'drv', 'cty',
       'hwy', 'fl', 'class'],
      dtype='object')

In [ ]:
mpg.

Display the summary statistics for the dataset.

In [145]:
mpg.describe()

displacement         year   cylinders        city     highway
count    234.000000   234.000000  234.000000  234.000000  234.000000
mean       3.471795  2003.500000    5.888889   16.858974   23.440171
std        1.291959     4.509646    1.611534    4.255946    5.954643
min        1.600000  1999.000000    4.000000    9.000000   12.000000
25%        2.400000  1999.000000    4.000000   14.000000   18.000000
50%        3.300000  2003.500000    6.000000   17.000000   24.000000
75%        4.600000  2008.000000    8.000000   19.000000   27.000000
max        7.000000  2008.000000    8.000000   35.000000   44.000000

In [142]:
mpg.rename(columns={'manufacturer': 'make', 'displ': 'displacement', 'cyl': 'cylinders', 'drv': 'drive', 'cty':'city', 'hwy': 'highway'}, inplace=True)

In [143]:
mpg

make   model  displacement  year  cylinders       trans drive  \
1          audi      a4           1.8  1999          4    auto(l5)     f   
2          audi      a4           1.8  1999          4  manual(m5)     f   
3          audi      a4           2.0  2008          4  manual(m6)     f   
4          audi      a4           2.0  2008          4    auto(av)     f   
5          audi      a4           2.8  1999          6    auto(l5)     f   
..          ...     ...           ...   ...        ...         ...   ...   
230  volkswagen  passat           2.0  2008          4    auto(s6)     f   
231  volkswagen  passat           2.0  2008          4  manual(m6)     f   
232  volkswagen  passat           2.8  1999          6    auto(l5)     f   
233  volkswagen  passat           2.8  1999          6  manual(m5)     f   
234  volkswagen  passat           3.6  2008          6    auto(s6)     f   

     city  highway fl    class  
1      18       29  p  compact  
2      21       29  p  compact  
3      20       31  p  compact  
4      21       30  p  compact  
5      16       26  p  compact  
..    ...      ... ..      ...  
230    19       28  p  midsize  
231    21       29  p  midsize  
232    16       26  p  midsize  
233    18       26  p  midsize  
234    17       26  p  midsize  

[234 rows x 11 columns]

How many different manufacturers are there? 15

In [159]:
len(mpg['make'].value_counts())

15

How many different models are there? 38

In [160]:
len(mpg['model'].value_counts())

38

Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [161]:
mpg['mileage_difference'] = mpg.highway - mpg.city

In [163]:
mpg.head(5)

make model  displacement  year  cylinders       trans drive  city  highway  \
1  audi    a4           1.8  1999          4    auto(l5)     f    18       29   
2  audi    a4           1.8  1999          4  manual(m5)     f    21       29   
3  audi    a4           2.0  2008          4  manual(m6)     f    20       31   
4  audi    a4           2.0  2008          4    auto(av)     f    21       30   
5  audi    a4           2.8  1999          6    auto(l5)     f    16       26   

  fl    class  mileage_difference  
1  p  compact                  11  
2  p  compact                   8  
3  p  compact                  11  
4  p  compact                   9  
5  p  compact                  10

Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [164]:
mpg['average_mileage'] = (mpg.highway + mpg.city) / 2

Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

In [166]:
mpg['is_automatic'] = (mpg['trans'].str.contains('auto'))

In [170]:
mpg['is_auto2'] = np.where(mpg.trans.str.contains('auto'), True, False)

In [173]:
mpg[['is_automatic', 'is_auto2']].head(5)

is_automatic  is_auto2
1          True      True
2         False     False
3         False     False
4          True      True
5          True      True

Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average? Honda

In [182]:
mpg.groupby(by='make').average_mileage.agg(['mean']).sort_values(by='mean', ascending=False)

mean
make                 
honda       28.500000
volkswagen  25.074074
hyundai     22.750000
subaru      22.428571
audi        22.027778
toyota      21.720588
pontiac     21.700000
nissan      21.346154
chevrolet   18.447368
ford        16.680000
mercury     15.625000
jeep        15.562500
dodge       15.540541
lincoln     14.166667
land rover  14.000000

Do automatic or manual cars have better miles per gallon? Manual

In [192]:
mpg.groupby(['is_auto2']).average_mileage.agg(['mean']).sort_values(by='mean')

mean
is_auto2           
True      19.130573
False     22.227273

## Exercises III

Use your get_db_url function to help you explore the data from the chipotle database.

In [196]:
url_chip = get_db_url(user, host, password, 'chipotle')

In [198]:
sql_tbl_orders = '''
SELECT * 
FROM orders;
'''

In [199]:
orders = pd.read_sql(sql_tbl_orders, url_chip)

In [200]:
orders.head()

id  order_id  quantity                              item_name  \
0   1         1         1           Chips and Fresh Tomato Salsa   
1   2         1         1                                   Izze   
2   3         1         1                       Nantucket Nectar   
3   4         1         1  Chips and Tomatillo-Green Chili Salsa   
4   5         2         2                           Chicken Bowl   

                                                           choice_description  \
0                                                                         nan   
1                                                                [Clementine]   
2                                                                     [Apple]   
3                                                                         nan   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   

  item_price  
0     $2.39   
1     $3.39   
2     $3.39   
3     $2.39   
4    $16.98

What is the total price for each order?

In [212]:
orders['int_price'] = orders['item_price'].str.strip('$').astype('float64')

In [213]:
orders.groupby(by='order_id').int_price.sum()

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: int_price, Length: 1834, dtype: float64

What are the most popular 3 items?

In [229]:
orders.groupby(by='item_name').quantity.sum().nlargest(3, keep='all')

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

Which item has produced the most revenue?

In [234]:
orders.groupby(by='item_name').int_price.sum().nlargest(1)

item_name
Chicken Bowl    7342.73
Name: int_price, dtype: float64

Join the employees and titles DataFrames together.

In [242]:
emp_titles = employees.merge(titles, how='inner', on='emp_no')

In [246]:
emp_titles.head(10)

emp_no  birth_date first_name  last_name gender   hire_date  \
0   10001  1953-09-02     Georgi    Facello      M  1986-06-26   
1   10002  1964-06-02    Bezalel     Simmel      F  1985-11-21   
2   10003  1959-12-03      Parto    Bamford      M  1986-08-28   
3   10004  1954-05-01  Chirstian    Koblick      M  1986-12-01   
4   10004  1954-05-01  Chirstian    Koblick      M  1986-12-01   
5   10005  1955-01-21    Kyoichi   Maliniak      M  1989-09-12   
6   10005  1955-01-21    Kyoichi   Maliniak      M  1989-09-12   
7   10006  1953-04-20     Anneke    Preusig      F  1989-06-02   
8   10007  1957-05-23    Tzvetan  Zielinski      F  1989-02-10   
9   10007  1957-05-23    Tzvetan  Zielinski      F  1989-02-10   

             title   from_date     to_date  
0  Senior Engineer  1986-06-26  9999-01-01  
1            Staff  1996-08-03  9999-01-01  
2  Senior Engineer  1995-12-03  9999-01-01  
3         Engineer  1986-12-01  1995-12-01  
4  Senior Engineer  1995-12-01  9999-01-01  
5     Senior Staff  1996-09-12  9999-01-01  
6            Staff  1989-09-12  1996-09-12  
7  Senior Engineer  1990-08-05  9999-01-01  
8     Senior Staff  1996-02-11  9999-01-01  
9            Staff  1989-02-10  1996-02-11

For each title, find the hire date of the employee that was hired most recently with that title.

In [265]:
#group by the title
#sort it by hired date
#return the most hired date
emp_titles.groupby(by='title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [271]:
sql_tbl_emp_ttle_dpt_dptno = '''
select * 
from employees
join titles USING (emp_no)
join dept_emp using (emp_no)
join departments using (dept_no)
'''

emp_titles_dept = pd.read_sql(sql_tbl_emp_ttle_dpt_dptno, url_emp)

In [274]:
emp_titles_dept.head(5)

dept_no  emp_no  birth_date first_name last_name gender   hire_date  \
0    d009   10011  1953-11-07       Mary     Sluis      F  1990-01-22   
1    d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
2    d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
3    d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   
4    d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   

          title   from_date     to_date   from_date     to_date  \
0         Staff  1990-01-22  1996-11-09  1990-01-22  1996-11-09   
1  Senior Staff  1996-09-20  9999-01-01  1989-09-20  9999-01-01   
2         Staff  1989-09-20  1996-09-20  1989-09-20  9999-01-01   
3  Senior Staff  2000-05-04  9999-01-01  1992-05-04  9999-01-01   
4         Staff  1992-05-04  2000-05-04  1992-05-04  9999-01-01   

          dept_name  
0  Customer Service  
1  Customer Service  
2  Customer Service  
3  Customer Service  
4  Customer Service

In [275]:
pd.crosstab(emp_titles_dept.title, emp_titles_dept.dept_name)

dept_name           Customer Service  Development  Finance  Human Resources  \
title                                                                         
Assistant Engineer               298         7769        0                0   
Engineer                        2362        58135        0                0   
Manager                            4            2        2                2   
Senior Engineer                 2027        49326        0                0   
Senior Staff                   13925         1247    12139            12274   
Staff                          16150         1424    13929            14342   
Technique Leader                 309         7683        0                0   

dept_name           Marketing  Production  Quality Management  Research  Sales  
title                                                                           
Assistant Engineer          0        6445                1831       378      0  
Engineer                    0       49649               13852      2986      0  
Manager                     2           4                   4         2      2  
Senior Engineer             0       42205               11864      2570      0  
Senior Staff            13940        1270                   0     11637  36191  
Staff                   16196        1478                   0     13495  41808  
Technique Leader            0        6557                1795       393      0